// pui-react-tabs/tabs.js
/*(c) Copyright 2015 Pivotal Software, Inc. All Rights Reserved.*/
'use strict';

var _extends = require('babel-runtime/helpers/extends')['default'];

var _objectWithoutProperties = require('babel-runtime/helpers/object-without-properties')['default'];

var _interopRequireDefault = require('babel-runtime/helpers/interop-require-default')['default'];

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactBootstrapLibTabs = require('react-bootstrap/lib/Tabs');

var _reactBootstrapLibTabs2 = _interopRequireDefault(_reactBootstrapLibTabs);

var _reactBootstrapLibTab = require('react-bootstrap/lib/Tab');

var _reactBootstrapLibTab2 = _interopRequireDefault(_reactBootstrapLibTab);

var _reactBootstrapLibAccordion = require('react-bootstrap/lib/Accordion');

var _reactBootstrapLibAccordion2 = _interopRequireDefault(_reactBootstrapLibAccordion);

var _reactBootstrapLibPanel = require('react-bootstrap/lib/Panel');

var _reactBootstrapLibPanel2 = _interopRequireDefault(_reactBootstrapLibPanel);

var _lodashUniqueid = require('lodash.uniqueid');

var _lodashUniqueid2 = _interopRequireDefault(_lodashUniqueid);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _raf = require('raf');

var _raf2 = _interopRequireDefault(_raf);

var _mediaSize = require('./media-size');

var _mediaSize2 = _interopRequireDefault(_mediaSize);

var BaseTabs = _react2['default'].createClass({
  displayName: 'BaseTabs',

  propTypes: {
    defaultActiveKey: _react2['default'].PropTypes.any,
    tabType: _react2['default'].PropTypes.oneOf(['tab-simple', 'tab-simple-alt', 'tab-left']),
    responsiveBreakpoint: _react2['default'].PropTypes.oneOf(['xs', 'sm', 'md', 'lg']),
    largeScreenClassName: _react2['default'].PropTypes.string,
    smallScreenClassName: _react2['default'].PropTypes.string,
    onSelect: _react2['default'].PropTypes.func,
    position: _react2['default'].PropTypes.oneOf(['top', 'left']),
    tabWidth: _react2['default'].PropTypes.number,
    paneWidth: _react2['default'].PropTypes.number,
    id: _react2['default'].PropTypes.string
  },

  getDefaultProps: function getDefaultProps() {
    return {
      responsiveBreakpoint: 'xs'
    };
  },

  getInitialState: function getInitialState() {
    return {
      activeKey: this.props.defaultActiveKey,
      smallScreen: false,
      id: _lodashUniqueid2['default']('pui-react-tabs-')
    };
  },

  setActiveKey: function setActiveKey(key) {
    var previousActiveKey = this.state.activeKey;
    this.setState({
      activeKey: key,
      previousActiveKey: previousActiveKey
    });
  },

  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    if (nextProps.defaultActiveKey !== this.props.defaultActiveKey) {
      this.setActiveKey(nextProps.defaultActiveKey);
    }
  },

  componentDidMount: function componentDidMount() {
    _raf2['default'](this.checkScreenSize);
  },

  checkScreenSize: function checkScreenSize() {
    if (!this.isMounted()) {
      return;
    } else {
      if (_mediaSize2['default'].matches(this.props.responsiveBreakpoint)) {
        this.setState({ smallScreen: false });
      } else {
        this.setState({ smallScreen: true });
      }

      _raf2['default'](this.checkScreenSize);
    }
  },

  handleSelect: function handleSelect(key) {
    if (!this.props.onSelect) {
      this.setActiveKey(key);
    } else {
      this.props.onSelect(key);
    }
  },

  render: function render() {
    var _props = this.props;
    var defaultActiveKey = _props.defaultActiveKey;
    var children = _props.children;
    var responsiveBreakpoint = _props.responsiveBreakpoint;
    var tabType = _props.tabType;
    var largeScreenClassName = _props.largeScreenClassName;
    var smallScreenClassName = _props.smallScreenClassName;
    var onSelect = _props.onSelect;
    var position = _props.position;
    var tabWidth = _props.tabWidth;
    var paneWidth = _props.paneWidth;

    var props = _objectWithoutProperties(_props, ['defaultActiveKey', 'children', 'responsiveBreakpoint', 'tabType', 'largeScreenClassName', 'smallScreenClassName', 'onSelect', 'position', 'tabWidth', 'paneWidth']);

    var largeScreenClasses = _classnames2['default']([tabType, largeScreenClassName]);
    var smallScreenClasses = _classnames2['default']([tabType + '-small-screen', smallScreenClassName]);

    var tabs = undefined;

    if (this.state.smallScreen) {
      var childrenAsPanels = _react2['default'].Children.map(children, function (child) {
        var _child$props = child.props;
        var title = _child$props.title;

        var childProps = _objectWithoutProperties(_child$props, ['title']);

        return _react2['default'].createElement(_reactBootstrapLibPanel2['default'], _extends({ header: title }, childProps));
      });

      tabs = _react2['default'].createElement(
        _reactBootstrapLibAccordion2['default'],
        _extends({ className: smallScreenClasses,
          activeKey: this.state.activeKey,
          onSelect: this.handleSelect
        }, props),
        childrenAsPanels
      );
    } else {
      tabs = _react2['default'].createElement(
        'div',
        { className: largeScreenClasses },
        _react2['default'].createElement(
          _reactBootstrapLibTabs2['default'],
          _extends({ position: position,
            tabWidth: tabWidth,
            paneWidth: paneWidth,
            activeKey: this.state.activeKey,
            onSelect: this.handleSelect
          }, _extends({ id: this.state.id }, props)),
          children
        )
      );
    }

    return _react2['default'].createElement(
      'div',
      null,
      tabs
    );
  }
});

/**
 * @component SimpleTabs
 * @description Tab-based content toggling
 *

 * @example ```js
 * var SimpleTabs = require('pui-react-tabs').SimpleTabs;
 * var Tab = require('pui-react-tabs').Tab;
 * var MyComponent = React.createClass({
 *   render() {
 *     return (
 *       <SimpleTabs defaultActiveKey={2}>
 *         <Tab eventKey={1}>Content for tab 1</Tab>
 *         <Tab eventKey={2}>Content for tab 2</Tab>
 *         <Tab eventKey={3}>Content for tab 3</Tab>
 *       </SimpleTabs>
 *     );
 *   }
 * });
 * ```
 *
 * @property responsiveBreakpoint {one of: `"xs"`, `"sm"`, `"md"`, `lg`} The
 * size at which the small-screen tabs (accordion-style) should switch to
 * large-screen tabs (folder-style)
 *
 * @property defaultActiveKey {should equal one of your tab's event keys} The
 * tab which will start out open
 *
 * @property smallScreenClassName {css class} Will be applied to small screen
 * tabs only
 *
 * @property largeScreenClassName {css class} Will be applied to large screen
 * tabs only
 *
 * @property onSelect {function} Will override default behavior when clicking
 * on a tab. If you want to retain the default behavior as well as add new
 * functionality, change default active key in the function you provide
 *
 */
var SimpleTabs = _react2['default'].createClass({
  displayName: 'SimpleTabs',

  render: function render() {
    return _react2['default'].createElement(BaseTabs, _extends({}, this.props, { tabType: 'tab-simple' }));
  }
});

/**
 * @component SimpleAltTabs
 * @description `<SimpleTabs>` with different styling
 *
 * @example ```js
 * var SimpleAltTabs = require('pui-react-tabs').SimpleAltTabs;
 * var Tab = require('pui-react-tabs').Tab;
 * var MyComponent = React.createClass({
 *   render() {
 *     return (
 *       <SimpleAltTabs defaultActiveKey={2}>
 *         <Tab eventKey={1}>Content for tab 1</Tab>
 *         <Tab eventKey={2}>Content for tab 2</Tab>
 *         <Tab eventKey={3}>Content for tab 3</Tab>
 *       </SimpleAltTabs>
 *     );
 *   }
 * });
 * ```
 *
 * @property responsiveBreakpoint {one of: `"xs"`, `"sm"`, `"md"`, `lg`} The
 * size at which the small-screen tabs (accordion-style) should switch to
 * large-screen tabs (folder-style)
 *
 * @property defaultActiveKey {should equal one of your tab's event keys} The
 * tab which will start out open
 *
 * @property smallScreenClassName {css class} Will be applied to small screen
 * tabs only
 *
 * @property largeScreenClassName {css class} Will be applied to large screen
 * tabs only
 *
 * @property onSelect {function} Will override default behavior when clicking
 * on a tab. If you want to retain the default behavior as well as add new
 * functionality, change default active key in the function you provide
 *
 */
var SimpleAltTabs = _react2['default'].createClass({
  displayName: 'SimpleAltTabs',

  render: function render() {
    return _react2['default'].createElement(BaseTabs, _extends({}, this.props, { tabType: 'tab-simple-alt' }));
  }
});

/**
 * @component LeftTabs
 * @description Tabs with the nav stacked on the left
 *
 * @example ```js
 * var LeftTabs = require('pui-react-tabs').LeftTabs;
 * var Tab = require('pui-react-tabs').Tab;
 * var MyComponent = React.createClass({
 *   render() {
 *     return (
 *       <LeftTabs defaultActiveKey={1} tabWidth={3} paneWidth={9}>
 *         <Tab eventKey={1} title="Tab 1">Wow!</Tab>
 *         <Tab eventKey={2} title="Tab 2">
 *           <h2>Neat!</h2>
 *           <span>So much content.</span>
 *         </Tab>
 *       </LeftTabs>
 *     );
 *   }
 * });
 * ```
 *
 * @property responsiveBreakpoint {one of: `"xs"`, `"sm"`, `"md"`, `lg`} The
 * size at which the small-screen tabs (accordion-style) should switch to
 * large-screen tabs (folder-style)
 *
 * @property defaultActiveKey {should equal one of your tab's event keys} The
 * tab which will start out open
 *
 * @property smallScreenClassName {css class} Will be applied to small screen
 * tabs only
 *
 * @property largeScreenClassName {css class} Will be applied to large screen
 * tabs only
 *
 * @property onSelect {function} Will override default behavior when clicking
 * on a tab. If you want to retain the default behavior as well as add new
 * functionality, change default active key in the function you provide
 *
 * @property tabWidth {number} Takes the number of bs Columns. Optional: the default is 6.
 *
 * @property paneWidth {number} Takes the number of bs Columns. Optional: the default is 24 - tabWidth.
 *
 */
var LeftTabs = _react2['default'].createClass({
  displayName: 'LeftTabs',

  propTypes: {
    position: _react2['default'].PropTypes.oneOf(['top', 'left']),
    tabWidth: _react2['default'].PropTypes.number,
    paneWidth: _react2['default'].PropTypes.number
  },
  getDefaultProps: function getDefaultProps() {
    return {
      position: 'left',
      tabWidth: 6
    };
  },
  render: function render() {
    var _props2 = this.props;
    var tabWidth = _props2.tabWidth;
    var paneWidth = _props2.paneWidth;

    var props = _objectWithoutProperties(_props2, ['tabWidth', 'paneWidth']);

    if (!paneWidth) {
      paneWidth = 24 - tabWidth;
    }
    return _react2['default'].createElement(BaseTabs, _extends({}, props, { tabWidth: tabWidth, paneWidth: paneWidth, tabType: 'tab-left' }));
  }
});

/**
 * @component Tab
 * @description A container for content in a `<SimpleTabs>` or `<SimpleAltTabs>`
 *
 * @property eventKey {Number} An identifier for the tab
 *
 */
var Tab = _reactBootstrapLibTab2['default'];

module.exports = {
  BaseTabs: BaseTabs,
  SimpleTabs: SimpleTabs,
  SimpleAltTabs: SimpleAltTabs,
  Tab: Tab,
  LeftTabs: LeftTabs
};
